File: 2021/python/day05.py
from aocUtil import *

def getAllCoords(line):
    coords = []
    points = line.split(" -> ")
    firstPoint = points[0].split(",")
    secondPoint = points[1].split(",")
    x1 = int(firstPoint[0])
    y1 = int(firstPoint[1])
    x2 = int(secondPoint[0])
    y2 = int(secondPoint[1])
    if x1 == x2:
        for i in range(min(y1, y2), max(y1, y2) + 1):
            coords.append(f"{x1},{i}")
    elif y1 == y2:
        for i in range(min(x1, x2), max(x1, x2) + 1):
            coords.append(f"{i},{y1}")
    else:
        diff = abs(x1 - x2)
        xNeg = -1 if x1 - x2 > 0 else 1
        yNeg = -1 if y1 - y2 > 0 else 1
        for i in range(0, diff + 1):
            coords.append(f"{x1 + (i * xNeg)},{y1 + (i * yNeg)}")
    
    return coords

def solve(inputs):
    vents = {}
    for line in inputs:
        for point in getAllCoords(line):
            vents[point] = vents.get(point, 0) + 1
    return len([x for x in vents.values() if x > 1])


if __name__ == "__main__":
    print(solve(getInputs("05")))

File: 2021/scala/adventofcode/src/main/scala/days/Day05.scala
import scala.annotation.tailrec

object Day05 {

    case class Points(x: List[Int], y: List[Int])

    def mapLineToPoints(line: String): Vector[(Int, Int)] = {
        val pointArray = line.split(" -> ").map(_.split(",").map(_.toInt)).transpose
        val points = Points(pointArray(0).toList, pointArray(1).toList)
        points match {
            // Vertical
            case points if points.x.forall(_ == points.x.head) => (points.y.min to points.y.max).map((points.x.head, _)).toVector

            // Horizontal
            case points if points.y.forall(_ == points.y.head) => (points.x.min to points.x.max).map((_, points.y.head)).toVector

            // Diagonal
            case _ => {
                val diff = points.x.reduce((x1, x2) => (x1 - x2).abs)
                val xNeg = if (points.x(0) - points.x(1) > 0) -1 else 1
                val yNeg = if (points.y(0) - points.y(1) > 0) -1 else 1
                (0 to diff).map(i => (points.x(0) + i * xNeg, points.y(0) + i * yNeg)).toVector
            }
        }
    }

    def getOverlappingPoints(inputs: Vector[String]) : Int = {
        inputs.flatMap(mapLineToPoints(_)).groupMapReduce(identity)(_ => 1)(_ + _).filter(_._2 >= 2).size
    }

    def solveDay: Int = {
        getOverlappingPoints(Util.getInput("05"))
    }
}
